libafl_bolts/src/llmp.rs

```diff
@@ -2266,38 +2272,46 @@ where
     where
         F: FnMut(ClientId, Tag, Flags, &[u8]) -> Result<LlmpMsgHookResult, Error>,
     {
-        #[cfg(feature = "std")]
-        let current_time = current_time();
         let mut new_messages = false;
         for i in 0..self.llmp_clients.len() {
             let client_id = self.llmp_clients[i].id;
             match unsafe { self.handle_new_msgs(client_id, on_new_msg) } {
                 Ok(has_messages) => {
-                    // See if we need to remove this client, in case no new messages got brokered, and it's not a listener
-                    #[cfg(feature = "std")]
-                    if !has_messages && !self.listeners.iter().any(|&x| x == client_id) {
-                        let last_msg_time = self.llmp_clients[i].last_msg_time;
-                        if last_msg_time < current_time
-                            && current_time - last_msg_time > self.client_timeout
-                        {
-                            self.clients_to_remove.push(i);
-                            #[cfg(feature = "llmp_debug")]
-                            log::info!("Client #{:#?} timed out. Removing.", client_id);
-                        }
-                    }
                     new_messages = has_messages;
                 }
-                Err(Error::ShuttingDown) => self.clients_to_remove.push(i),
+                Err(Error::ShuttingDown) => {
+                    self.clients_to_remove.insert(client_id);
+                }
                 Err(err) => return Err(err),
             }
         }
 
-        // After brokering, remove all clients we don't want to keep.
-        for i in self.clients_to_remove.iter().rev() {
-            let client_id = self.llmp_clients[*i].id;
-            log::info!("Client #{:#?} disconnected.", client_id);
-            self.llmp_clients.remove(*i);
+        let possible_remove = self.clients_to_remove.len();
```

Comment:
Like, _much_ slower

Reply:
I changed to set because i later discovered this could have duplicates.
there's two way that you can disconnect.
first is the client sends send_exiting()
second is the restarter sends detach_from_broker(). and we don't know which message can arrive earlier. it depends. if they arrive at the same time span then we have duplicates.

Reply:
adding to hashset is O(1) but if we use vec add checks for duplicates everytime we could have O(n^2) worstcase right?

Reply:
O doesn't matter here, accessing a vec is almost 0 overhead (one deref), accessing a set is slow as f*ck, it needs to hash the guy, look up some random linked list.
It'd probably need 16k entries to be faster with a set than a vec...

Reply:
Big O only matters if you're in Google scale - if you're close to the hardware vecs will basically always outperform Hash-anythings. Feel free to try it out with a `bench` or google around, but just thinking about it, the constant overhead, plus cache misses, etc., makes a _huge_ difference if the numbers of entries is small